### PR TITLE
Invariant F1, F2 and F4 fix

### PR DESCRIPTION
**Pull request by mattcushman**
# Description of change
## High level
* inflator-adjust reserve change in invariant settle
* Fixed F4 invariant

<!---
Add the `Status: Needs Auditor Approval` tags
CHANGES IN /SRC DIR:
- Divide t0Debt change from delta of reserves by the inflator
- Checked that there is zero deposit between a bucket and the next bucket found when searching for more deposit

# Description of bug or vulnerability and solution
* Invariant F1 was failing, because the invariant test we removing a different amount from the Fenwick Tree in settle when there was zero collateral remaining in loan and buckets were taking bad debt.
* This properly computes the t0 debt amount removed in the invariant settle function

This change should have no gas impact.  Size:
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,356B  (99.10%)
  ERC20Pool                -  23,802B  (96.85%)

Before:
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,040B  (97.82%)
  ERC20Pool                -  23,486B  (95.56%)

### CODE REVIEW

**Review by grandizzy — APPROVED** [review]
looks good, though I still get a F1 failure with sequence below

```
[FAIL. Reason: Incorrect deposits in bucket]
        [Sequence]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=bucketTake(uint256,uint256,bool,uint256), args=[115792089237316195423570985008687907853269984665640564039457584007913129639935, 6795291738248207466439618906053379568494097950800765921, true, 3]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=transferLps(uint256,uint256,uint256,uint256), args=[0, 2, 115792089237316195423570985008687907853269984665640564039457584007913129639934, 37656453947547]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=transferLps(uint256,uint256,uint256,uint256), args=[92622820366245905188853183146703854120022346672090660843674298451026172587106, 183446267711816861480078627892463, 2897678913945692926089276357210169, 74641623131693841147728]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=repayDebt(uint256,uint256), args=[0, 2]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=settleAuction(uint256,uint256,uint256), args=[2999999989933768610196238324134038719785353588, 4015, 143337254384252888583578505336]

 invariant_fenwick_depositAtIndex_F1() (runs: 1000, calls: 9995, reverts: 357)
```

and 

```
[FAIL. Reason: Incorrect deposits in bucket]
        [Sequence]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=settleAuction(uint256,uint256,uint256), args=[14915, 830, 2000000000000000000]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=removeQuoteToken(uint256,uint256,uint256), args=[5094, 427, 12621]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=removeQuoteToken(uint256,uint256,uint256), args=[4097, 7696, 893]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=bucketTake(uint256,uint256,bool,uint256), args=[2537099847, 9035629251980016, false, 0]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=kickAuction(uint256,uint256,uint256), args=[188907134199659870792852212814998950173773091668, 115792089237316195423570985008687907853269984665640564039457584007913129639935, 2821694]
                sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=settleAuction(uint256,uint256,uint256), args=[67, 20647, 91780542375707836935890807625238258332898240643225749252082938297103823675455]

 invariant_fenwick_depositAtIndex_F1() (runs: 1000, calls: 9995, reverts: 5)
```

**Comment by mattcushman** [comment]
> looks good, though I still get a F1 failure with sequence below
> 
> ```
> [FAIL. Reason: Incorrect deposits in bucket]
>         [Sequence]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=bucketTake(uint256,uint256,bool,uint256), args=[115792089237316195423570985008687907853269984665640564039457584007913129639935, 6795291738248207466439618906053379568494097950800765921, true, 3]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=transferLps(uint256,uint256,uint256,uint256), args=[0, 2, 115792089237316195423570985008687907853269984665640564039457584007913129639934, 37656453947547]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=transferLps(uint256,uint256,uint256,uint256), args=[92622820366245905188853183146703854120022346672090660843674298451026172587106, 183446267711816861480078627892463, 2897678913945692926089276357210169, 74641623131693841147728]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=repayDebt(uint256,uint256), args=[0, 2]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/ReservePoolHandler.sol:ReservePoolHandler]0x111893f9ebf6e485603808b7a8dcc3e1d3f854ad calldata=settleAuction(uint256,uint256,uint256), args=[2999999989933768610196238324134038719785353588, 4015, 143337254384252888583578505336]
> 
>  invariant_fenwick_depositAtIndex_F1() (runs: 1000, calls: 9995, reverts: 357)
> ```
> 
> and
> 
> ```
> [FAIL. Reason: Incorrect deposits in bucket]
>         [Sequence]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=settleAuction(uint256,uint256,uint256), args=[14915, 830, 2000000000000000000]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=removeQuoteToken(uint256,uint256,uint256), args=[5094, 427, 12621]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=removeQuoteToken(uint256,uint256,uint256), args=[4097, 7696, 893]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=bucketTake(uint256,uint256,bool,uint256), args=[2537099847, 9035629251980016, false, 0]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=kickAuction(uint256,uint256,uint256), args=[188907134199659870792852212814998950173773091668, 115792089237316195423570985008687907853269984665640564039457584007913129639935, 2821694]
>                 sender=0x0000000000000000000000000000000000001234 addr=[tests/forge/ERC20Pool/invariants/handlers/LiquidationPoolHandler.sol:LiquidationPoolHandler]0x32e03c6b1b446c7a4381852a82f7cd4beb00b17d calldata=settleAuction(uint256,uint256,uint256), args=[67, 20647, 91780542375707836935890807625238258332898240643225749252082938297103823675455]
> 
>  invariant_fenwick_depositAtIndex_F1() (runs: 1000, calls: 9995, reverts: 5)
> ```

yes, might still be some other causes causing similar fails elsewhere.  this should clear some up

File: src/base/Pool.sol [hunk 1]
```diff
@@ -749,6 +749,11 @@ abstract contract Pool is Clone, ReentrancyGuard, Multicall, IPool {
         );
     }
 
+    /// @inheritdoc IPoolDerivedState
+    function prefixSum(uint256 index_) external view returns (uint256) {
```
Comment (grandizzy) [inline src/base/Pool.sol]: Let's add override as we do for other functions. Can we come up with a more meaningful name in pool context? (Like we do for `depositSize()` or `depositIndex()`)

**Review by grandizzy — CHANGES_REQUESTED** [review]

File: src/base/Pool.sol (hunk 1 above)
Comment (grandizzy) [reply src/base/Pool.sol]: I added override with https://github.com/ajna-finance/contracts/pull/697/commits/3c88dcc0404dc4464d41f4101f219aa32737662d if we can come up with a better name for `prefixSum` frunction I'm all good to merge in develop, no more failures when test invariant with a depth of 10

Comment (mattcushman) [reply src/base/Pool.sol]: thank you, was figuring there was a better way to do this than just exposing it.  I don't care much about the name, how about `depositUpToIndex` ?

Comment (grandizzy) [reply src/base/Pool.sol]: sounds good!

**Review by prat-gpt — APPROVED** [review]
LGTM!

File: src/base/Pool.sol (hunk 1 above)
Comment (prat-gpt) [reply src/base/Pool.sol]: I have updated this here [88cf507](https://github.com/ajna-finance/contracts/pull/697/commits/88cf5076e4137c844435b0ee66bfbc729b44f2b0)

**Review by grandizzy — APPROVED** [review]

**Pull request merged**